websock/session.nim

```diff
@@ -116,13 +119,61 @@ proc send*(
     else:
       default(MaskKey)
 
-  if opcode in {Opcode.Text, Opcode.Cont, Opcode.Binary}:
-    await ws.writeMessage(
-      data, opcode, maskKey, ws.extensions)
+  if opcode.isControl:
+    await ws.writeControl(data, opcode, maskKey)
+  else:
+    await ws.writeMessage(data, opcode, maskKey, ws.extensions)
 
-    return
+proc doSend(
+  ws: WSSession,
+  data: seq[byte] = @[],
+  opcode: Opcode
+  ): Future[void] =
+  let
+    retFut = newFuture[void]("doSend")
+    sendFut = ws.nonCancellableSend(data, opcode)
+
+  proc handleSend {.async.} =
+    try:
+      await sendFut
+      retFut.complete()
+    except CatchableError as exc:
+      retFut.fail(exc)
+
+  asyncSpawn handleSend()
+  retFut
+
+proc sendLoop(ws: WSSession) {.gcsafe, async.} =
+  while ws.sendQueue.len > 0:
+    let task = ws.sendQueue.popFirst()
+    if task.fut.cancelled:
+      continue
```

Comment:
only checking for `cancelled` to behave similarly in case that client abuses API and completes/fails in advance, as if the queue was not used